### PR TITLE
remove the stringInterning feature flag from the compiler

### PR DESCRIPTION
**Pull request by paulbrauner-da**
Context: https://github.com/digital-asset/daml/issues/18240

And delete the resulting dead code paths from the encoder.

### CODE REVIEW

File: compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs [hunk 1]
```diff
@@ -566,7 +546,7 @@ encodeBuiltinExpr = \case
 
 encodeExpr' :: Expr -> Encode P.Expr
 encodeExpr' = \case
-    EVar v -> expr . either P.ExprSumVarStr P.ExprSumVarInternedStr <$> encodeName' unExprVarName v
+    EVar v -> expr . P.ExprSumVarInternedStr <$> encodeNameId unExprVarName v
```
Comment (dylant-da) [inline compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs:549]: It's nice to see all these places where the logic becomes simpler due to merging fallible encodeName' into infallible encodeNameId

File: compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs [hunk 2]
```diff
@@ -935,10 +913,9 @@ encodeDefDataType DefDataType{..} = do
             defDataType_FieldsFields <- encodeFieldsWithTypes unVariantConName fs
             pure $ P.DefDataTypeDataConsVariant P.DefDataType_Fields{..}
         DataEnum cs -> do
-            mangledAndInterned <- encodeNames (map unVariantConName cs)
-            let (defDataType_EnumConstructorsConstructorsStr, defDataType_EnumConstructorsConstructorsInternedStr) = case mangledAndInterned of
-                    Left mangled -> (V.fromList mangled, V.empty)
-                    Right mangledIds -> (V.empty, V.fromList mangledIds)
+            mangledIds <- encodeNames (map unVariantConName cs)
+            let defDataType_EnumConstructorsConstructorsStr = V.empty
```
Comment (dylant-da) [inline compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs:917]: May be worth making a TODO here for eventually removing the constructors_str field from proto, like e.g. on 115 of compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs above.

**Review by dylant-da — APPROVED** [review]
LGTM, thank you for this :D

File: compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs (hunk 1 above)
Comment (paulbrauner-da) [reply compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs:549]: Yes! This series of PRs to remove features is very satisfactory to work on :)

File: compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs (hunk 2 above)
Comment (paulbrauner-da) [reply compiler/daml-lf-proto-encode/src/DA/Daml/LF/Proto3/EncodeV2.hs:917]: Done

**Review by akrmn — APPROVED** [review]
nice cleanup, thank you @paulbrauner-da!

**Pull request merged**